temporal-sdk/src/main/java/io/temporal/internal/worker/EagerActivitySlotsReservation.java

```diff
@@ -66,34 +74,43 @@ public void applyToRequest(RespondWorkflowTaskCompletedRequest.Builder mutableRe
   public void handleResponse(RespondWorkflowTaskCompletedResponse serverResponse) {
     int activityTasksCount = serverResponse.getActivityTasksCount();
     Preconditions.checkArgument(
-        activityTasksCount <= this.outstandingReservationSlotsCount,
+        activityTasksCount <= this.reservedSlots.size(),
         "Unexpectedly received %s eager activities though we only requested %s",
         activityTasksCount,
-        this.outstandingReservationSlotsCount);
+        this.reservedSlots.size());
 
-    releaseSlots(this.outstandingReservationSlotsCount - activityTasksCount);
+    // Release any slots that we won't be using
+    releaseSlots(this.reservedSlots.size() - activityTasksCount);
 
     for (PollActivityTaskQueueResponse act : serverResponse.getActivityTasksList()) {
       // don't release slots here, instead the semaphore release reference is passed to the activity
       // worker to release when the activity is done
-      this.eagerActivityDispatcher.dispatchActivity(act);
+      SlotPermit permit = this.reservedSlots.remove(0);
+      this.eagerActivityDispatcher.dispatchActivity(act, permit);
     }
 
-    this.outstandingReservationSlotsCount = 0;
+    this.reservedSlots.clear();
   }
 
   @Override
   public void close() {
-    if (this.outstandingReservationSlotsCount > 0)
-      releaseSlots(this.outstandingReservationSlotsCount);
+    if (!this.reservedSlots.isEmpty()) {
+      releaseSlots(this.reservedSlots.size());
+    }
   }
 
   private void releaseSlots(int slotsToRelease) {
-    if (slotsToRelease > this.outstandingReservationSlotsCount)
+    if (slotsToRelease <= 0) return;
```

Comment:
If we throw on `slotsToRelease ` being to large shouldn't we throw on it being negative as well?